File: puzzle19.py
import re

file = open("input.txt", "r")
maze = []
for line in file:
    maze.append(line.rstrip('\n'))

x = 0
y = 0
direction = 'd'
for i in range (0, len(maze[0])):
    if maze[0][i] != ' ':
        x = i

letters = ''
steps = 0
while (1):
    steps += 1
    if direction == 'd':
        y += 1
        if maze[y][x] == '|':
            continue
        elif maze[y][x] == '+':
            if maze[y][x+1] != ' ' and maze[y][x+1] != '|':
                direction = 'r'
            else:
                direction = 'l'
        elif maze[y][x] == '-':
            continue
        elif maze[y][x] == ' ':
            break
        else:
            letters += maze[y][x]

    elif direction == 'u':
        y -= 1
        if maze[y][x] == '|':
            continue
        elif maze[y][x] == '+':
            if maze[y][x+1] != ' ' and maze[y][x+1] != '|':
                direction = 'r'
            else:
                direction = 'l'
        elif maze[y][x] == '-':
            continue
        elif maze[y][x] == ' ':
            break
        else:
            letters += maze[y][x]



    elif direction == 'r':
        x += 1
        if maze[y][x] == '-':
            continue
        elif maze[y][x] == '+':
            if maze[y+1][x] != ' ' and maze[y+1][x] != '-':
                direction = 'd'
            else:
                direction = 'u'
        elif maze[y][x] == '|':
            continue
        elif maze[y][x] == ' ':
            break
        else:
            letters += maze[y][x]


    elif direction == 'l':
        x -= 1
        if maze[y][x] == '-':
            continue
        elif maze[y][x] == '+':
            if maze[y+1][x] != ' ' and maze[y+1][x] != '-':
                direction = 'd'
            else:
                direction = 'u'
        elif maze[y][x] == '|':
            continue
        elif maze[y][x] == ' ':
            break
        else:
            letters += maze[y][x]
print "sol 1: " + letters
print "sol 2: " + str(steps)

File: 2020/day19/puzzle19.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	lines := readInputLines("input.txt")
	//	lines := readInputLines("../sample.txt")
	length := 132
	//	length = 31

	var rule []string
	var text []string
	// List was sorted
	for i, line := range lines {
		if i >= length {
			text = append(text, line)
		} else {
			rule = append(rule, line)
		}
	}

	regex := findRegex(rule)
	re := regexp.MustCompile("^" + regex + "$")
	answer1 := 0
	for _, i := range text {
		matched := re.MatchString(i)
		if matched {
			answer1++
		}
	}
	answer2 := 0

	for loop := 0; loop < 10; loop++ {
		newRule := loopWithDepth(rule, loop)
		regex = findRegex(newRule)
		re = regexp.MustCompile("^" + regex + "$")
		answer2 = 0
		for _, i := range text {
			matched := re.MatchString(i)
			if matched {
				answer2++
			}
		}
	}

	fmt.Println(fmt.Sprintf("Answer 1 : %d", answer1))
	fmt.Println(fmt.Sprintf("Answer 2 : %d", answer2))
}

func loopWithDepth(rules []string, loop int) []string {
	var newRule []string
	for _, rule := range rules {
		splitNum := strings.Split(rule, ":")
		if splitNum[0] == "8" {
			newRule = append(newRule, rule+" + ")
			continue
		}
		if splitNum[0] == "11" {
			eleven := "11: 42 31 "
			for j := 2; j < loop; j++ {
				eleven += " | "
				for i := 0; i < j; i++ {
					eleven += " 42 "
				}
				for i := 0; i < j; i++ {
					eleven += " 31 "
				}
			}
			newRule = append(newRule, eleven)
			continue
		}
		newRule = append(newRule, rule)
	}

	return newRule

}

func findRegex(input []string) string {
	rules := make(map[int]string)
	for {
		if _, ok := rules[0]; ok {
			break
		}
		for _, rule := range input {
			goNext := false
			ruleSplit := strings.Split(rule, ": ")
			i := s2i(ruleSplit[0])
			if _, ok := rules[i]; ok {
				continue
			}
			parts := strings.Fields(rule)
			parts = parts[1:]

			if string(parts[0][0]) == "\"" {
				rules[i] = string(parts[0][1])
				continue
			}
			regex := "(?:"
			for _, part := range parts {
				if part == "|" {
					regex += "|"
				} else if part == "+" {
					regex += "+"
				} else {
					token := s2i(part)
					a, ok := rules[token]
					if !ok {
						goNext = true
						break
					}
					regex += a
				}
			}
			if goNext {
				continue
			}
			regex += ")"
			rules[i] = regex
		}
	}
	return rules[0]
}

func s2i(val string) int {
	num, err := strconv.Atoi(val)
	if err != nil {
		fmt.Println("OH NO! OH NO! NOT AN INT!")
		fmt.Println(err)
	}
	return num
}

func readInputLines(fileName string) []string {
	file, err := os.Open(fileName)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	scanner.Split(bufio.ScanLines)
	var text []string

	for scanner.Scan() {
		text = append(text, scanner.Text())
	}

	return text
}
